### PR TITLE
Proposing new structure of DALI examples

### PR DESCRIPTION
**Pull request by szalpal**
Signed-off-by: Michał Szołucha <mszolucha@nvidia.com>

This PR proposes a new structure of examples in DALI. The main idea here is that the tutorials web page structure shall reflect directory structure. The proposed structure can be seen [HERE](https://github.com/szalpal/DALI/tree/examples_refactor/docs/examples)

### CODE REVIEW

**Comment by szalpal** [comment]
!build

File: docs/examples/README.md [hunk 1]
```diff
@@ -1,17 +1,14 @@
 # Examples
 
-This directory contains examples of DALI usage. List of examples:
+This directory contains examples of DALI usage.
```
Comment (klecki) [inline docs/examples/README.md]: We should make this readme appear in the proper docs instead of having it only on GH. But it can be done as a followup. I would also list what the `General` means.

File: docs/examples/frameworks/mxnet/mxnet-external_input.ipynb [hunk 2]
```diff
@@ -47,7 +47,7 @@
    "source": [
     "class ExternalInputIterator(object):\n",
     "    def __init__(self, batch_size, device_id, num_gpus):\n",
-    "        self.images_dir = \"../images/\"\n",
+    "        self.images_dir = \"../../images/\"\n",
```
Comment (JanuszL) [inline docs/examples/frameworks/mxnet/mxnet-external_input.ipynb]:
```suggestion
    "        self.images_dir = \"../data/images/\"\n",
```
?

File: docs/examples/frameworks/mxnet/mxnet-various-readers.ipynb [hunk 3]
```diff
@@ -44,7 +44,7 @@
     "lmdb_folder = os.path.join(test_data_root, 'db', 'lmdb')\n",
     "\n",
     "# image dir with plain jpeg files\n",
-    "image_dir = \"../images\"\n",
+    "image_dir = \"../../images\"\n",
```
Comment (JanuszL) [inline docs/examples/frameworks/mxnet/mxnet-various-readers.ipynb]:
```suggestion
    "image_dir = \"../data/images\"\n",
```
?

File: docs/examples/frameworks/pytorch/pytorch-external_input.ipynb [hunk 4]
```diff
@@ -47,7 +47,7 @@
    "source": [
     "class ExternalInputIterator(object):\n",
     "    def __init__(self, batch_size, device_id, num_gpus):\n",
-    "        self.images_dir = \"../images/\"\n",
+    "        self.images_dir = \"../../images/\"\n",
```
Comment (JanuszL) [inline docs/examples/frameworks/pytorch/pytorch-external_input.ipynb]: As above

File: docs/examples/frameworks/pytorch/pytorch-various-readers.ipynb [hunk 5]
```diff
@@ -44,7 +44,7 @@
     "lmdb_folder = os.path.join(test_data_root, 'db', 'lmdb')\n",
     "\n",
     "# image dir with plain jpeg files\n",
-    "image_dir = \"../images\"\n",
+    "image_dir = \"../../images\"\n",
```
Comment (JanuszL) [inline docs/examples/frameworks/pytorch/pytorch-various-readers.ipynb]: As above

File: docs/examples/frameworks/tensorflow/tensorflow-various-readers.ipynb [hunk 6]
```diff
@@ -44,7 +44,7 @@
     "lmdb_folder = os.path.join(test_data_root, 'db', 'lmdb')\n",
     "\n",
     "# image dir with plain jpeg files\n",
-    "image_dir = \"../images\"\n",
+    "image_dir = \"../../images\"\n",
```
Comment (JanuszL) [inline docs/examples/frameworks/tensorflow/tensorflow-various-readers.ipynb]: As above

File: docs/examples/general/external_input.ipynb [hunk 7]
```diff
@@ -45,7 +45,7 @@
    "source": [
     "class ExternalInputIterator(object):\n",
     "    def __init__(self, batch_size):\n",
-    "        self.images_dir = \"images/\"\n",
+    "        self.images_dir = \"../images/\"\n",
```
Comment (JanuszL) [inline docs/examples/general/external_input.ipynb]: As above

File: docs/examples/general/index.rst [hunk 8]
```diff
@@ -0,0 +1,13 @@
+General
+=======
+
+.. toctree::
+   :maxdepth: 2
+
+   custom_operator/create_a_custom_operator.ipynb
```
Comment (JanuszL) [inline docs/examples/general/index.rst]: Maybe custom operator and python_operator should go into separate category "Extending DALI"?

File: docs/examples/general/index.rst [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+   expressions/index
+   external_input.ipynb
```
Comment (JanuszL) [inline docs/examples/general/index.rst]: External input should go to data loading

File: docs/examples/general/multigpu.ipynb [hunk 10]
```diff
@@ -30,7 +30,7 @@
     "import nvidia.dali.types as types\n",
     "from nvidia.dali.pipeline import Pipeline\n",
     "\n",
-    "image_dir = \"images\"\n",
+    "image_dir = \"../images\"\n",
```
Comment (JanuszL) [inline docs/examples/general/multigpu.ipynb]: As above

File: docs/examples/image_processing/brightness_contrast_example.ipynb [hunk 11]
```diff
@@ -28,7 +28,7 @@
     "import matplotlib.pyplot as plt\n",
     "\n",
     "batch_size = 1\n",
-    "image_filename = \"images\""
+    "image_filename = \"../images\""
```
Comment (JanuszL) [inline docs/examples/image_processing/brightness_contrast_example.ipynb]: As above

File: docs/examples/image_processing/color_space_conversion.ipynb [hunk 12]
```diff
@@ -27,7 +27,7 @@
     "import nvidia.dali.types as types\n",
     "from nvidia.dali.pipeline import Pipeline\n",
     "\n",
-    "image_dir = \"images\"\n",
+    "image_dir = \"../images\"\n",
```
Comment (JanuszL) [inline docs/examples/image_processing/color_space_conversion.ipynb]: As above

File: docs/examples/image_processing/decoder_examples.ipynb [hunk 13]
```diff
@@ -42,7 +42,7 @@
     "        plt.axis(\"off\")\n",
     "        plt.imshow(image_batch.at(j))\n",
     "\n",
-    "image_dir = \"images\"\n",
+    "image_dir = \"../images\"\n",
```
Comment (JanuszL) [inline docs/examples/image_processing/decoder_examples.ipynb]: As above

File: docs/examples/image_processing/hsv_example.ipynb [hunk 14]
```diff
@@ -30,7 +30,7 @@
     "import matplotlib.pyplot as plt\n",
     "\n",
     "batch_size = 1\n",
-    "image_filename = \"images\""
+    "image_filename = \"../images\""
```
Comment (JanuszL) [inline docs/examples/image_processing/hsv_example.ipynb]: As above

File: docs/examples/image_processing/index.rst [hunk 15]
```diff
@@ -0,0 +1,15 @@
+Image Processing
+================
+
+.. toctree::
+   :maxdepth: 2
+
+   data_loading/index
```
Comment (JanuszL) [inline docs/examples/image_processing/index.rst]: I don't think that data loading should go there. Yes, we use images as an examples, but it doesn't make this example relevant only to this particular use case.

File: docs/examples/image_processing/index.rst [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+   augmentation_gallery.ipynb
+   brightness_contrast_example.ipynb
+   color_space_conversion.ipynb
+   decoder_examples.ipynb
+   hsv_example.ipynb
+   warp_3D.ipynb
```
Comment (JanuszL) [inline docs/examples/image_processing/index.rst]: How about a separate sub-category for 3D?

File: docs/examples/sequence_processing/index.rst [hunk 17]
```diff
@@ -0,0 +1,9 @@
+Sequence Processing
+===================
+
+.. toctree::
+   :maxdepth: 2
+
+   video/video_reader_simple_example.ipynb
```
Comment (JanuszL) [inline docs/examples/sequence_processing/index.rst]: I would rather put it into data loading. As not processing is done in this video examples.

File: docs/examples/sequence_processing/index.rst [hunk 18]
```diff
@@ -0,0 +1,9 @@
...
+   video/video_reader_label_example.ipynb
+   sequence_reader_simple_example.ipynb
```
Comment (JanuszL) [inline docs/examples/sequence_processing/index.rst]: Same here

File: docs/examples/use_cases/index.rst [hunk 19]
```diff
@@ -0,0 +1,9 @@
+Use Cases
+=========
+
+.. toctree::
+   :maxdepth: 2
+
+   detection_pipeline.ipynb
+   optical_flow_example.ipynb
```
Comment (JanuszL) [inline docs/examples/use_cases/index.rst]: I would put this into video category

File: docs/examples/use_cases/index.rst [hunk 20]
```diff
@@ -0,0 +1,9 @@
+Use Cases
+=========
+
+.. toctree::
+   :maxdepth: 2
+
+   detection_pipeline.ipynb
```
Comment (JanuszL) [inline docs/examples/use_cases/index.rst]: This is a vision example.

File: docs/examples/README.md [hunk 21]
```diff
@@ -1,17 +1,14 @@
...
 - [Getting started](getting%20started.ipynb) - build your first pipeline with DALI
-- [Augmentation gallery](augmentation_gallery.ipynb) - showcases different augmentations
-- [Color space conversion](color_space_conversion.ipynb) - showcases different color space conversion operations (e.g. RGB to YCbCr)
-- [LMDB data loading](dataloading_lmdb.ipynb) - example of reading data from LMDB database
-- [TFRecord data loading](dataloading_tfrecord.ipynb) - example of reading data from TensorFlow's TFRecord file
-- [RecordIO data loading](dataloading_recordio.ipynb) - example of reading data from MXNet's RecordIO file
-- [Serialization](serialization.ipynb) - example of pipeline serialization
+- [General](general) - General topics about using DALI, regardless if you want to do image, audio or any other processing.
+- [Image Processing](image_processing), [Sequence Processing](sequence_processing), [Audio Processing](audio_processing) - Tutorials, that present topics specific to the type of processed data
+- [Use Cases](use_cases) - Take a look on some real-life scenarios, where DALI can proove to be useful.
```
Comment (JanuszL) [inline docs/examples/README.md]: I think this is an artificial category that tries to group examples form different categories.

File: docs/examples/audio_processing/audio_decoder.ipynb [hunk 22]
```diff
@@ -25,7 +25,7 @@
     "import numpy as np\n",
     "\n",
     "batch_size = 1\n",
-    "audio_files = \"audio\""
+    "audio_files = \"../audio\""
```
Comment (JanuszL) [inline docs/examples/audio_processing/audio_decoder.ipynb]:
```suggestion
    "audio_files = \"../data/audio\""
```
?

File: docs/examples/frameworks/index.rst [hunk 23]
```diff
@@ -0,0 +1,9 @@
+Sequence Processing
```
Comment (JanuszL) [inline docs/examples/frameworks/index.rst]: ?

File: docs/examples/frameworks/index.rst [hunk 24]
```diff
@@ -0,0 +1,9 @@
...
+===================
+
+.. toctree::
+   :maxdepth: 2
+
+   video/video_reader_simple_example.ipynb
```
Comment (JanuszL) [inline docs/examples/frameworks/index.rst]: I would put all of this into video category. Not the frameworks.
Also I don't see any rst that gathers MXNet, TF and Pytorch examples.

File: docs/examples/frameworks/index.rst (hunk 23 above)
Comment (szalpal) [reply docs/examples/frameworks/index.rst]: My bad, removed this file.

File: docs/examples/frameworks/index.rst (hunk 24 above)
Comment (szalpal) [reply docs/examples/frameworks/index.rst]: My bad, removed this file.

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

File: docs/examples/general/index.rst (hunk 8 above)
Comment (klecki) [reply docs/examples/general/index.rst]: :+1:

**Comment by szalpal** [comment]
!build

**Comment by jantonguirao** [comment]
@szalpal  Why did we rename `audio` and `sequence` directories to `audio_processing` and `sequence_processing` ??

**Comment by szalpal** [comment]
> @szalpal Why did we rename `audio` and `sequence` directories to `audio_processing` and `sequence_processing` ??

You comment answers it ;)

In `audio` there was a mix of audio data and audio examples. As a similarity, there was `image` directory, only with image data. This way it's clear where is what.

File: docs/examples/sequence_processing/index.rst (hunk 17 above)
Comment (szalpal) [reply docs/examples/sequence_processing/index.rst]: That's why I wanted to put `data_loading` in `image` category. When using DALI you need to know how to load your data properly for your particular use-case...

I's leave it here, thinking about a guy that wants to now, how to process video in DALI. He'd go to `video` first, not `general` or any other stuff...

File: docs/examples/image_processing/index.rst (hunk 16 above)
Comment (szalpal) [reply docs/examples/image_processing/index.rst]: Since there's only 1 example for it, I'd pass right now. In future sure, but now it'll be just empty...

File: docs/examples/image_processing/index.rst (hunk 15 above)
Comment (szalpal) [reply docs/examples/image_processing/index.rst]: Done

File: docs/examples/use_cases/index.rst (hunk 20 above)
Comment (szalpal) [reply docs/examples/use_cases/index.rst]: This detection pipeline is basically what we use in SSD, so I figured it's more of a use-case. Even disregarding the SSD, this is basically a use-case, as there's more to it than only image processing (e.g. labels passing and stuff).

File: docs/examples/README.md (hunk 21 above)
Comment (szalpal) [reply docs/examples/README.md]: We can put there, e.g. pipelines for SSD, audio pipelines used in some models, real-life scenarios, this kind of stuff.

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

File: docs/_templates/layout.html [hunk 25]
```diff
@@ -48,4 +48,5 @@
     border-top: solid 3px rgba(59,93,0,1);
   }
   </style>
+  <script src="//assets.adobedtm.com/b92787824f2e0e9b68dc2e993f9bd995339fe417/satelliteLib-7ba51e58dc61bcb0e9311aadd02a0108ab24cc6c.js"></script>
```
Comment (JanuszL) [inline docs/_templates/layout.html]: It is no longer here. Rebase.

File: docs/examples/README.md (hunk 21 above)
Comment (JanuszL) [reply docs/examples/README.md]: pipelines for SSD -> mage Processing
audio pipelines -> Audio Processing
Still it doesn't make sense to put them in to one bag when we have this fine grind separation somewhere else.

File: docs/plugins/mxnet_tutorials.rst [hunk 26]
```diff
@@ -5,5 +5,5 @@ MXNet
    :maxdepth: 2
    :glob:
 
-   ../examples/mxnet/**
+   ../examples/frameworks/mxnet/index
```
Comment (JanuszL) [inline docs/plugins/mxnet_tutorials.rst]: There was an idea to put API first and then examples. What do you think?

File: qa/TL0_python_self_test_frameworks/test_cupy.sh [hunk 27]
```diff
@@ -4,7 +4,10 @@ pip_packages="nose numpy cupy"
 target_dir=./dali/test/python
 
 test_body() {
-    nosetests --verbose -m '(?:^|[\b_\./-])[Tt]est.*cupy' test_dltensor_operator.py
+    if [[ ${PYTHON_VERSION} != 2.7 ]]
```
Comment (JanuszL) [inline qa/TL0_python_self_test_frameworks/test_cupy.sh]: It is no longer here. Please rebase.

File: docs/examples/sequence_processing/index.rst (hunk 17 above)
Comment (JanuszL) [reply docs/examples/sequence_processing/index.rst]: If he would go to video why it is inside `Sequence Processing`?

File: docs/examples/use_cases/index.rst (hunk 20 above)
Comment (JanuszL) [reply docs/examples/use_cases/index.rst]: Following that logic any complete pipeline is a use case - RN50, SSD, SuperRes...
I still think that `Use Cases` is a vague idea that could only confuse people.

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by klecki** [comment]
We need to place Multi GPU tutorial visible out of the box in `General`

**Comment by JanuszL** [comment]
I still think:
- External input should go to data loading
- ResNet50 examples and SSD could go to use cases but still I would link them inside particular FW section too as those examples are specific to FW as well

**Comment by szalpal** [comment]
> I still think:
> 
>     * External input should go to data loading
> 
>     * ResNet50 examples and SSD could go to use cases but still I would link them inside particular FW section too as those examples are specific to FW as well

I transfered `ExternalInput` to `data_loading`.

As for the latter, I'd prefer to remain the way it is now. One of the sins of previous examples structure way that directory structure didn't reflect the toctree. While ofc it's not neccessary, I'd say it's one of the reasons there was a mess. I'd prefer to keep it aligned.
And why RN50 and SSD in `use_cases`, not in `frameworks`? It's because they aren't simple-written, easy-to-understand examples of using DALI with given framework. On the contrary, the are pretty complicated and I feel that they belong more to use-cases than to DALI+FW tutorial.

If you insist I'd change it, but I'd prefer to keep it this way.

**Comment by szalpal** [comment]
!build

**Comment by JanuszL** [comment]
> > I still think:
> > ```
> > * External input should go to data loading
> > 
> > * ResNet50 examples and SSD could go to use cases but still I would link them inside particular FW section too as those examples are specific to FW as well
> > ```
> 
> I transfered `ExternalInput` to `data_loading`.
> 
> As for the latter, I'd prefer to remain the way it is now. One of the sins of previous examples structure way that directory structure didn't reflect the toctree. While ofc it's not neccessary, I'd say it's one of the reasons there was a mess. I'd prefer to keep it aligned.
> And why RN50 and SSD in `use_cases`, not in `frameworks`? It's because they aren't simple-written, easy-to-understand examples of using DALI with given framework. On the contrary, the are pretty complicated and I feel that they belong more to use-cases than to DALI+FW tutorial.
> 
> If you insist I'd change it, but I'd prefer to keep it this way.

OK

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build
